### PR TITLE
Wrapped Min and Max Functions

### PR DESCRIPTION
**Pull request by ShikharJ**
@isuruf Can you tell me why are the current builds failing? Everything seems to be in compliance.

### CODE REVIEW

**Comment by isuruf** [comment]
See this error message,

```
Error compiling Cython file:
------------------------------------------------------------
...
    cdef symengine.vec_basic v
    cdef Basic e_
    for e in args:
        e_ = sympify(e)
        v.push_back(e_.thisptr)
    return c2py(symengine.max(v))
                              ^
------------------------------------------------------------
symengine_wrapper.pyx:2250:31: Cannot assign type 'vec_basic' to 'RCP[const Basic]'
```

**Comment by ShikharJ** [comment]
I mean, isn't `symengine.max` supposed to take a `vec_basic`, or do I have to run a loop for individual elements?

File: symengine/lib/symengine.pxd [hunk 1]
```diff
@@ -417,6 +419,7 @@ cdef extern from "<symengine/functions.h>" namespace "SymEngine":
     cdef RCP[const Basic] acoth(RCP[const Basic] &arg) nogil except+
     cdef RCP[const Basic] function_symbol(string name, const vec_basic &arg) nogil except+
     cdef RCP[const Basic] abs(RCP[const Basic] &arg) nogil except+
+    cdef RCP[const Basic] max(RCP[const Basic] &arg) nogil except+
```
Comment (isuruf) [inline symengine/lib/symengine.pxd]: This function signature is wrong

**Comment by ShikharJ** [comment]
I think this is ready for review now.

**Comment by ShikharJ** [comment]
ping @isuruf

File: symengine/tests/test_functions.py [hunk 2]
```diff
@@ -73,6 +73,20 @@ def test_abs():
     assert abs(Integer(5)/3+x) != Integer(5)/3
     assert abs(Integer(5)/3+x) == abs(Integer(5)/3+x)
 
+def test_max():
+
+    assert max({Integer(6)/3, 1}) == 2
+    assert max({-2, 2}) == 2
+    assert max({2, 2}) == 2
+    assert max({0.2, 0.3}) == 0.3
```
Comment (isuruf) [inline symengine/tests/test_functions.py]: This is using python's `min` and `max`.

File: symengine/lib/symengine_wrapper.pyx [hunk 3]
```diff
@@ -2216,6 +2267,22 @@ def log(x, y = None):
...
 
+def max(*args):
+    cdef symengine.vec_basic v
+    cdef Basic e_
+    for e in args:
+        e_ = sympify(e)
+        v.push_back(e_.thisptr)
+    return c2py(symengine.max(v))
+
+def min(*args):
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: It's not good to change a function in the global namespace of python.  What does sympy do?

**Comment by ShikharJ** [comment]
@isuruf From what I observe, `Sympy` has a `Max` and `Min` class defined which inherits from `MaxMinBase`, which itself inherits from `Expr` and `LatticeOp`. Individual `max` and `min` functions are defined, though only as member functions, that too of some specific classes.

**Comment by ShikharJ** [comment]
@isuruf What do you suggest we do of the namespace issue?

**Comment by isuruf** [comment]
Same as in SymPy. We define the classes Max and Min only.

**Comment by ShikharJ** [comment]
@isuruf I'am facing trouble implementing that. `Min` and `Max` both have constructors of the order `Class::Class(const vec_basic &&args)` which is raising issues in the builds. Can you please take a look and suggest improvements?

**Comment by ShikharJ** [comment]
ping @isuruf

**Comment by isuruf** [comment]
Use `min` and `max`, don't use the constructors of `Max` and `Min`. Constructors of `Max` and `Min` are supposed to be used only when you know the result is a `Max` or a `Min`. You don't want to create a `Max(1, 2)`

**Comment by ShikharJ** [comment]
@isuruf I'm not sure what you mean. Wouldn't using `max` and `min` bring us back to the same problem of namespace?

**Comment by ShikharJ** [comment]
Ping @isuruf

File: symengine/tests/test_functions.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 from symengine import Symbol, sin, cos, sqrt, Add, Mul, function_symbol, Integer, log, E, symbols
-from symengine.lib.symengine_wrapper import Subs, Derivative
+from symengine.lib.symengine_wrapper import Subs, Derivative, _max, _min
```
Comment (isuruf) [inline symengine/tests/test_functions.py]: We don't want `_max` to be the public interface.
`Max` can be the public interface, but for that we need to rename the existing class to `_Max` and make `Max`, a pure python class and override `__new__` to return the value, like in https://github.com/symengine/symengine.py/blob/master/symengine/sympy_compat.py#L53-L54

Comment (ShikharJ) [inline symengine/tests/test_functions.py]: Thanks for clearing me up on this.

**Comment by ShikharJ** [comment]
@isuruf Can you please take a look now?

**Comment by ShikharJ** [comment]
@isuruf I have added a string to address the import issues. I'd like to raise an issue for a possible workaround for the above. Can you please take a look here?

**Comment by ShikharJ** [comment]
Ping @isuruf

**Comment by isuruf** [comment]
Shifting the tests to sympy_compat is not a real solution. we need to give the user a way to use `Max` and `Min` when they do `from symengine import *`

**Comment by ShikharJ** [comment]
@isuruf Shifting the tests was never intended as a solution to the import issue. It was more along the lines of current necessity. As I'm out of ideas regarding the solution of the namespace issue, I would like to raise an issue, for possible ideas regarding the same, after this is merged. I pinged for a final review of this code.

**Comment by ShikharJ** [comment]
+1

**Comment by ShikharJ** [comment]
Thanks for guiding me through @isuruf.

**Pull request merged**